File: src/services/excelExporter.js
import { RESULT_COLORS, VISIBLE_EXPORT_COLUMNS } from "./matcher.js";

function createCellStyle({ fillColor = "FFFFFF", bold = false, center = false } = {}) {
  return {
    font: { bold },
    fill: {
      patternType: "solid",
      fgColor: { rgb: fillColor },
    },
    alignment: {
      vertical: "center",
      horizontal: center ? "center" : "left",
      wrapText: true,
    },
    border: {
      top: { style: "thin", color: { rgb: "000000" } },
      bottom: { style: "thin", color: { rgb: "000000" } },
      left: { style: "thin", color: { rgb: "000000" } },
      right: { style: "thin", color: { rgb: "000000" } },
    },
  };
}

export async function exportResults({ semana, data }) {
  if (!data?.length) {
    throw new Error("No hay resultados para exportar.");
  }

  const xlsxModule = await import("xlsx-js-style");
  const XLSX = xlsxModule.default ?? xlsxModule;
  const cleanWeek = semana.trim();
  const title = `SEMANA ${cleanWeek}`;
  const headers = VISIBLE_EXPORT_COLUMNS.map(([, label]) => label);
  const rows = data.map((row) =>
    VISIBLE_EXPORT_COLUMNS.map(([fieldName]) => row[fieldName] ?? ""),
  );

  const worksheet = XLSX.utils.aoa_to_sheet([[title], headers, ...rows]);
  worksheet["!merges"] = [
    {
      s: { r: 0, c: 0 },
      e: { r: 0, c: VISIBLE_EXPORT_COLUMNS.length - 1 },
    },
  ];
  worksheet["!cols"] = VISIBLE_EXPORT_COLUMNS.map(([fieldName], index) => {
    const headerLength = headers[index].length;
    const maxValueLength = data.reduce((maxLength, row) => {
      const currentLength = String(row[fieldName] ?? "").length;
      return Math.max(maxLength, currentLength);
    }, headerLength);

    return { wch: Math.min(Math.max(maxValueLength + 2, 14), 45) };
  });

  const titleAddress = XLSX.utils.encode_cell({ r: 0, c: 0 });
  worksheet[titleAddress].s = {
    font: { bold: true, sz: 14 },
    alignment: { horizontal: "center", vertical: "center" },
  };

  headers.forEach((_, columnIndex) => {
    const headerAddress = XLSX.utils.encode_cell({ r: 1, c: columnIndex });
    worksheet[headerAddress].s = createCellStyle({
      fillColor: "D9EAD3",
      bold: true,
      center: true,
    });
  });

  data.forEach((row, rowIndex) => {
    const fillColor = RESULT_COLORS[row.resultado_cruce] ?? "FFFFFF";
    VISIBLE_EXPORT_COLUMNS.forEach((_, columnIndex) => {
      const cellAddress = XLSX.utils.encode_cell({ r: rowIndex + 2, c: columnIndex });
      if (worksheet[cellAddress]) {
        worksheet[cellAddress].s = createCellStyle({ fillColor });
      }
    });
  });

  const workbook = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(workbook, worksheet, "Resultado");
  XLSX.writeFile(workbook, `reporte_cruce_semana_${cleanWeek}.xlsx`);
}

File: backend/services/excel_exporter.py
from __future__ import annotations

from io import BytesIO

from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side

from .constants import RESULT_COLORS, VISIBLE_EXPORT_COLUMNS


def build_export_file(semana: str, rows: list[dict[str, str]]) -> BytesIO:
    workbook = Workbook()
    worksheet = workbook.active
    worksheet.title = "Resultado"

    last_column_index = len(VISIBLE_EXPORT_COLUMNS)
    worksheet.merge_cells(start_row=1, start_column=1, end_row=1, end_column=last_column_index)

    title_cell = worksheet.cell(row=1, column=1, value=f"SEMANA {semana}")
    title_cell.font = Font(bold=True, size=14)
    title_cell.alignment = Alignment(horizontal="center", vertical="center")

    header_fill = PatternFill(fill_type="solid", fgColor="D9EAD3")
    header_font = Font(bold=True)
    border = Border(
        left=Side(style="thin", color="000000"),
        right=Side(style="thin", color="000000"),
        top=Side(style="thin", color="000000"),
        bottom=Side(style="thin", color="000000"),
    )

    header_row = 2
    data_start_row = 3

    for index, (_, header_label) in enumerate(VISIBLE_EXPORT_COLUMNS, start=1):
        cell = worksheet.cell(row=header_row, column=index, value=header_label)
        cell.fill = header_fill
        cell.font = header_font
        cell.alignment = Alignment(horizontal="center", vertical="center")
        cell.border = border

    for row_index, row in enumerate(rows, start=data_start_row):
        result_color = RESULT_COLORS.get(row.get("resultado_cruce", ""), "FFFFFF")
        data_fill = PatternFill(fill_type="solid", fgColor=result_color)

        for column_index, (field_name, _) in enumerate(VISIBLE_EXPORT_COLUMNS, start=1):
            cell = worksheet.cell(row=row_index, column=column_index, value=row.get(field_name, ""))
            cell.fill = data_fill
            cell.border = border
            if field_name == "fecha":
                cell.alignment = Alignment(horizontal="center", vertical="center")
            else:
                cell.alignment = Alignment(vertical="top")

    for column_index, (field_name, header_label) in enumerate(VISIBLE_EXPORT_COLUMNS, start=1):
        max_length = len(header_label)
        for row in rows:
            max_length = max(max_length, len(str(row.get(field_name, "") or "")))
        worksheet.column_dimensions[worksheet.cell(row=header_row, column=column_index).column_letter].width = min(
            max(max_length + 2, 14),
            45,
        )

    worksheet.freeze_panes = "A3"

    output = BytesIO()
    workbook.save(output)
    output.seek(0)
    return output
